public/app/core/components/NestedFolderPicker/NestedFolderPicker.tsx

```diff
@@ -123,39 +117,105 @@ export function NestedFolderPicker({ value, onChange }: NestedFolderPickerProps)
     return result;
   }, [search, searchState.value, rootFoldersState.loading, rootFoldersState.value, folderOpenState, childrenForUID]);
 
-  const handleSelectionChange = useCallback(
-    (event: React.FormEvent<HTMLInputElement>, item: DashboardViewItem) => {
-      if (onChange) {
-        onChange({
-          uid: item.uid,
-          title: item.title,
-        });
-      }
-      setOverlayOpen(false);
-    },
-    [onChange]
-  );
-
   const { getTooltipProps, setTooltipRef, setTriggerRef, visible, triggerRef } = usePopperTooltip({
     visible: overlayOpen,
     placement: 'bottom',
     interactive: true,
     offset: [0, 0],
     trigger: 'click',
     onVisibleChange: (value: boolean) => {
-      // ensure search state is clean on opening the overlay
+      // ensure state is clean on opening the overlay
       if (value) {
         setSearch('');
+        setFocusedItemIndex(-1);
+        setAutoFocusButton(true);
       }
       setOverlayOpen(value);
     },
   });
 
+  const handleFolderExpand = useCallback(async (uid: string, newOpenState: boolean) => {
+    setFolderOpenState((old) => ({ ...old, [uid]: newOpenState }));
+
+    if (newOpenState) {
+      const folders = await listFolders(uid, undefined, 1, PAGE_SIZE);
+      setChildrenForUID((old) => ({ ...old, [uid]: folders }));
+    }
+  }, []);
+
+  const handleFolderSelect = useCallback(
+    (item: DashboardViewItem) => {
+      if (onChange) {
+        onChange({
+          uid: item.uid,
+          title: item.title,
+        });
+      }
+      setOverlayOpen(false);
+    },
+    [onChange]
+  );
+
   const isLoading = rootFoldersState.loading || searchState.loading;
   const error = rootFoldersState.error || searchState.error;
 
   const tree = flatTree;
 
+  const handleKeyDown = useCallback(
+    (ev: React.KeyboardEvent<HTMLInputElement>) => {
+      const foldersAreOpenable = !(search && searchState.value);
```

Comment:
lets hoist this up and reuse it here and down where we pass it into NestedFolderList